scripts/publish.js

```diff
@@ -0,0 +1,227 @@
+import { existsSync, readFileSync } from "node:fs";
+import { join } from "node:path";
+
+//-----------------------------------------------------------------------------
+// Read CLI Args
+//-----------------------------------------------------------------------------
+
+const dryRun = process.argv.includes("--dry-run");
+
+// for dry runs only output to console and don't execute anything
+const exec = dryRun ? text => console.log(text) : execSync;
+
+//-----------------------------------------------------------------------------
+// Helpers
+//-----------------------------------------------------------------------------
+
+/**
+ * Converts a GitHub Actions step output name to its corresponding environment variable name
+ * @param {string} stepId The ID of the step
+ * @param {string} outputName The name of the output
+ * @returns {string} The environment variable name
+ */
+function convertOutputToEnvVar(stepId, outputName) {
+	// Convert step ID to uppercase and replace hyphens with underscores
+	const normalizedStepId = stepId.toUpperCase().replace(/-/gu, "_");
+
+	// Convert output name to uppercase and replace slashes with double underscores
+	const normalizedOutputName = outputName
+		.toUpperCase()
+		.replace(/\//gu, "__")
+		.replace(/-/gu, "_");
+
+	return `STEPS_${normalizedStepId}_OUTPUTS_${normalizedOutputName}`;
+}
+
+/**
+ * Gets the output of a GitHub Actions step from the environment variables.
+ * @param {string} packageDir The directory of the package.
+ * @param {string} name The name of the output.
+ * @return {string} The output value.
+ */
+function getReleaseOutput(packageDir, name) {
+	return process.env[
+		convertOutputToEnvVar("release", `${packageDir}--${name}`)
+	];
+}
+
+/**
+ * Gets the list of packages to publish.
+ * @param {Array<string>} packageDirs The list of package directories.
+ * @return {Array<string>} The list of packages to publish.
+ */
+function getPackagesToPublish(packageDirs) {
+	return packageDirs.filter(
+		packageDir =>
+			getReleaseOutput(packageDir, "release_created") === "true",
+	);
+}
+
+/**
+ * Publishes the packages to npm. If one package fails to publish, the rest
+ * will still be published.
+ * @param {Array<string>} packageDirs The list of package directories.
+ * @return {Map<string,string>} A map of package directory to whether it was published successfully.
+ */
+function publishPackagesToNpm(packageDirs) {
+	console.log(
+		`Publishing packages to npm in this order: ${packageDirs.join(", ")}`,
+	);
+
+	const results = new Map();
+
+	for (const packageDir of packageDirs) {
+		console.log(`Publishing ${packageDir}...`);
+		try {
+			exec(`npm publish -w ${packageDir} --provenance`, {
+				stdio: "inherit",
+				env: process.env,
+			});
+
+			results.set(packageDir, "ok");
+		} catch (error) {
+			console.error(`Failed to publish ${packageDir} to npm`);
+			console.log(error.message);
+
+			results.set(packageDir, error.message);
+		}
```

Comment:
If the publishing of one package fails, the rest will continue? If that's the case, it might not always be what we want, since the rest of the packages might have the failed one's new version (that hasn't been published due to failure) in package.json dependencies.

Reply:
That's a good point. I suppose I could re-check the results of each package's dependencies before attempting to publish.

My goal here is to publish as many of the packages a possible because not every package depends on every other package.